psutil/arch/osx/cpu.c

```diff
@@ -109,7 +114,78 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
     );
 }
 
+#if defined(__arm64__) || defined(__aarch64__)
+PyObject *
+psutil_cpu_freq(PyObject *self, PyObject *args) {
+    uint32_t min;
+    uint32_t curr;
+    uint32_t pMin;
+    uint32_t eMin;
+    uint32_t max;
+    kern_return_t status;
+
+    CFDictionaryRef matching = IOServiceMatching("AppleARMIODevice");
+    if (matching == 0) {
+        return PyErr_Format(PyExc_RuntimeError, "IOServiceMatching call failed, 'AppleARMIODevice' not found");
+    }
+    io_iterator_t iter;
+
+    status = IOServiceGetMatchingServices(kIOMainPortDefault, matching, &iter);
+    if (status != KERN_SUCCESS) {
+        return PyErr_Format(PyExc_RuntimeError, "IOServiceGetMatchingServices call failed");
+    }
+
+    io_registry_entry_t entry;
+    while ((entry = IOIteratorNext(iter))) {
+        io_name_t name;
+        status = IORegistryEntryGetName(entry, name);
+        if (status != KERN_SUCCESS) {
+            continue;
+        }
+        if (strcmp(name, "pmgr") == 0) {
+            break;
+        }
+    }
+
+    if (entry == 0) {
+        return PyErr_Format(PyExc_RuntimeError, "'pmgr' entry was not found in AppleARMIODevice service");
+    }
+
+    CFTypeRef pCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states5-sram"), kCFAllocatorDefault, 0);
+    if (!pCoreRef) {
+        return PyErr_Format(PyExc_RuntimeError, "'voltage-states5-sram' property not found");
+    }
+
+    CFTypeRef eCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states1-sram"), kCFAllocatorDefault, 0);
+    if (!eCoreRef) {
+        return PyErr_Format(PyExc_RuntimeError, "'voltage-states1-sram' property not found");
```

Comment:
You need to release `pCoreRef` before returning. More in general, I suggest to centralize the cleanup logic in one place, and use a goto.
This construct is used pretty much everywhere in psutil, so you can search for "error:" in the C files to know what I mean, but long story short what I'm suggesting is something like this (not tested).
Every time there is an error, anywhere, just set exception and then do `goto error;`.

```c
PyObject *
psutil_cpu_freq(PyObject *self, PyObject *args) {
    CFTypeRef pCoreRef = NULL;
    CFTypeRef eCoreRef = NULL;
    io_iterator_t iter = NULL;
    io_registry_entry_t entry = NULL;

    ...

    eCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states1-sram"), kCFAllocatorDefault, 0);
    if (!eCoreRef) {
        PyErr_Format(PyExc_RuntimeError, "'voltage-states1-sram' property not found");
        goto error;
    }

    ...

    error:
        if (pCoreRef != NULL):
            CFRelease(pCoreRef);
        if (eCoreRef != NULL):
            CFRelease(eCoreRef);
        if (iter != NULL)
            IOObjectRelease(iter);
        if (entry != NULL)
            IOObjectRelease(entry);
        return NULL;
}
```